File: ts-src/py-src/minicode/tui/input_parser.py
from __future__ import annotations
import re
from dataclasses import dataclass
from typing import Union, Literal

# Pre-compiled regexes for escape sequence parsing
_SGR_MOUSE_RE = re.compile(r'^\x1b\[<(\d+);(\d+);(\d+)([Mm])')
_CSI_CURSOR_RE = re.compile(r'^\x1b\[(?:1;(\d+))?([A-DF-H])')
_CSI_TILDE_RE = re.compile(r'^\x1b\[(\d+)(?:;(\d+))?~')
_SS3_RE = re.compile(r'^\x1bO([A-DF-H])')
_ESC_CHAR_RE = re.compile(r'^\x1b([^\x1b\[O])')

ParsedKeyName = Literal[
    'return', 'tab', 'backspace', 'delete',
    'up', 'down', 'left', 'right',
    'pageup', 'pagedown', 'home', 'end', 'escape'
]

@dataclass(frozen=True)
class KeyEvent:
    name: str  # ParsedKeyName or 'a', 'c', etc. for ctrl-keys
    ctrl: bool
    meta: bool
    kind: str = "key"

@dataclass(frozen=True)
class TextEvent:
    text: str
    ctrl: bool
    meta: bool
    kind: str = "text"

@dataclass(frozen=True)
class WheelEvent:
    direction: Literal['up', 'down']
    kind: str = "wheel"

ParsedInputEvent = Union[KeyEvent, TextEvent, WheelEvent]

@dataclass(frozen=True)
class ParseResult:
    events: list[ParsedInputEvent]
    rest: str

CTRL_CHAR_TO_NAME: dict[str, str] = {
    '\x01': 'a',
    '\x03': 'c',
    '\x05': 'e',
    '\x0e': 'n',
    '\x0f': 'o',
    '\x10': 'p',
    '\x15': 'u',
}

def maybe_need_more_for_escape_sequence(chunk: str) -> bool:
    if not chunk:
        return False
    if chunk[0] != '\x1b':
        return False
    if len(chunk) == 1:
        return True
    
    # CSI
    if chunk[1] == '[':
        # SGR Mouse: ESC[<button;x;yM/m
        if len(chunk) >= 3 and chunk[2] == '<':
            return not any(c in 'Mm' for c in chunk[3:])
        # Legacy Mouse: ESC[M...
        if len(chunk) >= 3 and chunk[2] == 'M':
            return len(chunk) < 6
        # CSI cursor/tilde: look for terminator char (A-Z, a-z, ~)
        # Only digits, semicolons, and '?' are valid intermediate/parameter bytes
        for i in range(2, len(chunk)):
            c = chunk[i]
            if 'A' <= c <= 'Z' or 'a' <= c <= 'z' or c == '~':
                return False
            if c not in '0123456789;?':
                # Invalid character in CSI — not a valid sequence, stop waiting
                return False
        # All chars so far are parameter bytes, still waiting for terminator
        return True
        
    # SS3
    if chunk[1] == 'O':
        return len(chunk) < 3
        
    # ESC + char (Alt+char)
    # We already checked len(chunk) == 1. For Alt+char, 2 chars is complete.
    return False

def parse_escape_sequence(chunk: str) -> tuple[ParsedInputEvent | None, int]:
    if not chunk or chunk[0] != '\x1b':
        return None, 0
        
    if len(chunk) == 1:
        return KeyEvent(name='escape', ctrl=False, meta=False), 1

    # SGR Mouse: ESC[<button;x;yM/m
    sgr_match = _SGR_MOUSE_RE.match(chunk)
    if sgr_match:
        button = int(sgr_match.group(1))
        # wheel events (button & 0x43 == 0x40 → up, 0x41 → down)
        if (button & 0x43) == 0x40:
            return WheelEvent(direction='up'), sgr_match.end()
        elif (button & 0x43) == 0x41:
            return WheelEvent(direction='down'), sgr_match.end()
        return None, sgr_match.end()

    # Legacy mouse: ESC[M...
    if chunk.startswith('\x1b[M') and len(chunk) >= 6:
        button = ord(chunk[3])
        if (button & 0x43) == 0x40:
            return WheelEvent(direction='up'), 6
        elif (button & 0x43) == 0x41:
            return WheelEvent(direction='down'), 6
        return None, 6

    # CSI cursor: ESC[{1;modifier}A/B/C/D/H/F
    csi_cursor_match = _CSI_CURSOR_RE.match(chunk)
    if csi_cursor_match:
        mod_str = csi_cursor_match.group(1)
        key_char = csi_cursor_match.group(2)
        mod = int(mod_str) if mod_str else 1
        
        # Modifier logic: 2=Shift, 3=Alt, 4=Shift+Alt, 5=Ctrl, 6=Shift+Ctrl, 7=Alt+Ctrl, 8=Shift+Alt+Ctrl
        # 1=None. 
        # (mod - 1) & 4 -> Ctrl
        # (mod - 1) & 2 -> Alt/Meta
        ctrl = bool((mod - 1) & 4)
        meta = bool((mod - 1) & 2)
        
        name_map: dict[str, str] = {
            'A': 'up', 'B': 'down', 'C': 'right', 'D': 'left', 'H': 'home', 'F': 'end'
        }
        return KeyEvent(name=name_map[key_char], ctrl=ctrl, meta=meta), csi_cursor_match.end()

    # CSI tilde: ESC[N~ or ESC[N;M~ (with modifier)
    csi_tilde_match = _CSI_TILDE_RE.match(chunk)
    if csi_tilde_match:
        n = int(csi_tilde_match.group(1))
        mod_str = csi_tilde_match.group(2)
        mod = int(mod_str) if mod_str else 1
        ctrl = bool((mod - 1) & 4)
        meta = bool((mod - 1) & 2)
        # 1=home,3=delete,4=end,5=pageup,6=pagedown,7=home,8=end
        tilde_map: dict[int, str] = {
            1: 'home', 3: 'delete', 4: 'end', 5: 'pageup', 6: 'pagedown', 7: 'home', 8: 'end'
        }
        if n in tilde_map:
            return KeyEvent(name=tilde_map[n], ctrl=ctrl, meta=meta), csi_tilde_match.end()
        return None, csi_tilde_match.end()

    # SS3: ESC O A/B/C/D/H/F
    ss3_match = _SS3_RE.match(chunk)
    if ss3_match:
        key_char = ss3_match.group(1)
        name_map: dict[str, str] = {
            'A': 'up', 'B': 'down', 'C': 'right', 'D': 'left', 'H': 'home', 'F': 'end'
        }
        return KeyEvent(name=name_map[key_char], ctrl=False, meta=False), ss3_match.end()

    # ESC+Tab
    if chunk.startswith('\x1b\t'):
        return KeyEvent(name='tab', ctrl=False, meta=True), 2

    # ESC+char (Alt+char)
    esc_char_match = _ESC_CHAR_RE.match(chunk)
    if esc_char_match:
        char = esc_char_match.group(1)
        return TextEvent(text=char, ctrl=False, meta=True), 2
        
    # Default to bare escape if nothing else matches and we are not waiting for more
    return KeyEvent(name='escape', ctrl=False, meta=False), 1

def parse_input_chunk(chunk: str) -> ParseResult:
    events: list[ParsedInputEvent] = []
    i = 0
    while i < len(chunk):
        if maybe_need_more_for_escape_sequence(chunk[i:]):
            break

        char = chunk[i]

        # Escape sequence
        if char == '\x1b':
            event, consumed = parse_escape_sequence(chunk[i:])
            if event:
                events.append(event)
            i += consumed
            continue

        # CR, LF, CR+LF -> return
        if char == '\r':
            if i + 1 < len(chunk) and chunk[i+1] == '\n':
                i += 2
            else:
                i += 1
            events.append(KeyEvent(name='return', ctrl=False, meta=False))
            continue

        if char == '\n':
            events.append(KeyEvent(name='return', ctrl=False, meta=False))
            i += 1
            continue

        # Tab
        if char == '\t':
            events.append(KeyEvent(name='tab', ctrl=False, meta=False))
            i += 1
            continue

        # Backspace (0x7f, 0x08)
        if char in ('\x7f', '\x08'):
            events.append(KeyEvent(name='backspace', ctrl=False, meta=False))
            i += 1
            continue

        # Ctrl chars (0x01-0x1a)
        if '\x01' <= char <= '\x1a':
            if char in CTRL_CHAR_TO_NAME:
                events.append(KeyEvent(name=CTRL_CHAR_TO_NAME[char], ctrl=True, meta=False))
            # Swallow other control characters
            i += 1
            continue

        # Regular text
        events.append(TextEvent(text=char, ctrl=False, meta=False))
        i += 1

    return ParseResult(events=events, rest=chunk[i:])

File: ts-src/src/tui/input-parser.ts
type ParsedKeyName =
  | 'return'
  | 'tab'
  | 'backspace'
  | 'delete'
  | 'up'
  | 'down'
  | 'left'
  | 'right'
  | 'pageup'
  | 'pagedown'
  | 'home'
  | 'end'
  | 'escape'

type ParsedInputEvent =
  | {
      kind: 'key'
      name: ParsedKeyName
      ctrl: boolean
      meta: boolean
    }
  | {
      kind: 'text'
      text: string
      ctrl: boolean
      meta: boolean
    }
  | {
      kind: 'wheel'
      direction: 'up' | 'down'
    }

type ParseResult = {
  events: ParsedInputEvent[]
  rest: string
}

const ESC = '\u001b'
const CTRL_CHAR_TO_NAME: Record<string, string> = {
  '\u0001': 'a',
  '\u0003': 'c',
  '\u0005': 'e',
  '\u000e': 'n',
  '\u000f': 'o',
  '\u0010': 'p',
  '\u0015': 'u',
}

function maybeNeedMoreForEscapeSequence(input: string): boolean {
  if (!input.startsWith(ESC)) return false

  if (input === ESC) return true
  // Wait only for known-incomplete CSI prefixes.
  // If an unexpected character appears (for example "\u001b[\r"), do not
  // block the parser forever; let parseEscapeSequence fall back.
  if (input === '\u001b[') return true
  if (/^\u001b\[[<\d;?]*$/.test(input)) {
    return true
  }
  if (input.startsWith('\u001bO') && input.length < 3) {
    return true
  }

  return false
}

function parseEscapeSequence(input: string): {
  event: ParsedInputEvent | null
  length: number
} | null {
  let match = /^\u001b\[<(\d+);(\d+);(\d+)([Mm])/.exec(input)
  if (match) {
    const button = Number(match[1])
    const length = match[0].length
    if ((button & 0x43) === 0x40) {
      return { event: { kind: 'wheel', direction: 'up' }, length }
    }
    if ((button & 0x43) === 0x41) {
      return { event: { kind: 'wheel', direction: 'down' }, length }
    }
    return { event: null, length }
  }

  if (/^\u001b\[M.../.test(input)) {
    const seq = input.slice(0, 6)
    const button = seq.charCodeAt(3) - 32
    if ((button & 0x43) === 0x40) {
      return { event: { kind: 'wheel', direction: 'up' }, length: 6 }
    }
    if ((button & 0x43) === 0x41) {
      return { event: { kind: 'wheel', direction: 'down' }, length: 6 }
    }
    return { event: null, length: 6 }
  }

  match = /^\u001b\[(?:1;(\d+))?([ABCDHF])/.exec(input)
  if (match) {
    const modifier = Number(match[1] ?? '1')
    const meta = modifier === 3
    const ctrl = modifier === 5
    const nameMap: Record<string, ParsedKeyName> = {
      A: 'up',
      B: 'down',
      C: 'right',
      D: 'left',
      H: 'home',
      F: 'end',
    }
    return {
      event: {
        kind: 'key',
        name: nameMap[match[2]],
        ctrl,
        meta,
      },
      length: match[0].length,
    }
  }

  match = /^\u001b\[(\d+)~/.exec(input)
  if (match) {
    const nameMap: Record<string, ParsedKeyName> = {
      '1': 'home',
      '3': 'delete',
      '4': 'end',
      '5': 'pageup',
      '6': 'pagedown',
      '7': 'home',
      '8': 'end',
    }
    const name = nameMap[match[1]]
    if (!name) return { event: null, length: match[0].length }
    return {
      event: { kind: 'key', name, ctrl: false, meta: false },
      length: match[0].length,
    }
  }

  match = /^\u001bO([ABCDHF])/.exec(input)
  if (match) {
    const nameMap: Record<string, ParsedKeyName> = {
      A: 'up',
      B: 'down',
      C: 'right',
      D: 'left',
      H: 'home',
      F: 'end',
    }
    return {
      event: {
        kind: 'key',
        name: nameMap[match[1]],
        ctrl: false,
        meta: false,
      },
      length: match[0].length,
    }
  }

  if (input.startsWith('\u001b\t')) {
    return {
      event: { kind: 'key', name: 'tab', ctrl: false, meta: true },
      length: 2,
    }
  }

  if (input.length >= 2) {
    const char = input[1]
    if (char !== '[' && char !== 'O') {
      return {
        event: { kind: 'text', text: char, ctrl: false, meta: true },
        length: 2,
      }
    }
  }

  return {
    event: { kind: 'key', name: 'escape', ctrl: false, meta: false },
    length: 1,
  }
}

export function parseInputChunk(
  previousRest: string,
  chunk: Buffer | string,
): ParseResult {
  const input = previousRest + String(chunk)
  const events: ParsedInputEvent[] = []
  let index = 0

  while (index < input.length) {
    const remaining = input.slice(index)

    if (/^\[<\d+;\d+;\d+[Mm]/.test(remaining)) {
      index += remaining.match(/^\[<\d+;\d+;\d+[Mm]/)?.[0].length ?? 1
      continue
    }

    if (remaining[0] === ESC) {
      if (maybeNeedMoreForEscapeSequence(remaining)) {
        return { events, rest: remaining }
      }

      const parsed = parseEscapeSequence(remaining)
      if (parsed) {
        if (parsed.event) events.push(parsed.event)
        index += parsed.length
        continue
      }
    }

    const char = remaining[0]
    if (!char) break

    if (char === '\r' || char === '\n') {
      events.push({ kind: 'key', name: 'return', ctrl: false, meta: false })
      if (
        (char === '\r' && remaining[1] === '\n') ||
        (char === '\n' && remaining[1] === '\r')
      ) {
        index += 2
      } else {
        index += 1
      }
      continue
    }

    if (char === '\t') {
      events.push({ kind: 'key', name: 'tab', ctrl: false, meta: false })
      index += 1
      continue
    }

    if (char === '\u007f' || char === '\b') {
      events.push({ kind: 'key', name: 'backspace', ctrl: false, meta: false })
      index += 1
      continue
    }

    if (char >= '\u0001' && char <= '\u001a') {
      const name = CTRL_CHAR_TO_NAME[char]
      if (name) {
        events.push({ kind: 'text', text: name, ctrl: true, meta: false })
      }
      index += 1
      continue
    }

    if (char < ' ') {
      index += 1
      continue
    }

    events.push({ kind: 'text', text: char, ctrl: false, meta: false })
    index += 1
  }

  return { events, rest: '' }
}

export type { ParsedInputEvent }
